File: tools_java/tools/src/tools/Chapter.java
package tools;


import org.w3c.dom.Document;
import org.w3c.dom.Element;
/**
 * 
 * @author broihier
 * <pre>
 * Class for creating chapter (section/category) objects
 * </pre>
 */

public class Chapter {
		private Document doc = null;
		private String status = "UNTESTED";
		private String title;
		/** 
		 * Constructor for creating chapter objects
		 * 
		 * <pre>
		 * Pseudo-code:
		 * {@code
		 * create a chapter table of contents DOM object;
		 * insert the title and description into the DOM;
		 * foreach test case in the test category (chapter) {
		 *     create a test case object;
		 *     add a table of contents entry to the DOM based on the status of the test case object;
		 *     update the overall status of the chapter;
		 * } 
		 * write out the table of content DOM for this chapter;
		 * }
		 * </pre>
		 * @param category name that maps to 'category'.testDbCategoryTitle/Description files
		 * @param database Test database file that contains all of the unformatted document information
		 * 
		 */

		public Chapter (String category, TestDatabase database) {
			DOMizeTemplate DOMizedTemplate = new DOMizeTemplate("../tools_configuration/TOC_Template.html","./"+category+"_toc.html"); 
			this.doc = DOMizedTemplate.getDoc();
			// insert section title into test case DOM
			Element element = Utilities.findElementById(doc,"title");
			this.title = database.getCategoryTitle();
			element.appendChild(doc.createTextNode(database.getCategoryTitle()));
			
			// insert section description into test case DOM
			element = Utilities.findElementById(doc,"description");
			element.appendChild(doc.createTextNode(database.getCategoryDescriptionText()));
			
			// insert test cases
			
			element = Utilities.findElementById(doc,"list");
			for (String cases : database.getIdentificationIDs()) {
				String status = (new TestCase(cases,category,database)).getStatus();
				Element listItem = doc.createElement("li");
				Element anchor = doc.createElement("a");
				anchor.setAttribute("href", category+"/"+cases+".html");
				if (status.equals("PASSED")) {
					anchor.setAttribute("style", "color: 00F000");
				} else if (status.equals("FAILED")) {
					anchor.setAttribute("style", "color: FF0000");
				}
				anchor.appendChild(doc.createTextNode(cases + " - " + database.getIdentificationText(cases)));
				listItem.appendChild(anchor);
				element.appendChild(listItem);
				updateChapterStatus(status);
			}
			
			// write TOC DOM
			DOMizedTemplate.writeDocument();
		}

		/** 
		 * Method to return the overall status of the chapter
		 * 
		 * @return status - getter
		 * 
		 */
		public String getStatus() {
			return status;
		}

		/** 
		 * Method to return the title of the chapter
		 * 
		 * @return title - getter
		 * 
		 */
		public String getTitle() {
			return title;
		}
		/** 
		 * Private method to determine the overall state of the chapter based on individual test case results
		 * 
		 */
		private void updateChapterStatus (String status) {
			if (status.equals("UNTESTED")) {
				if (! this.status.equals("UNTESTED")){
					this.status = "NOTDONE";
				}
			} else {
				if (status.equals("FAILED")) {
					this.status = "FAILED";
				} else {
					if (this.status.equals("UNTESTED")){
						this.status = "PASSED";
					}
				}
			}
		}
}

File: tools_C++/src/Chapter.cpp
/*
 * Chapter.cpp
 *
 *  Created on: Apr 15, 2017
 *      Author: broihier
 */

#include "Chapter.h"
#include "TestDatabase.h"
#include "DOMizeTemplate.h"
#include "TestCase.h"
#include <iostream>
#include <stdio.h>


Chapter::Chapter(std::string category, TestDatabase& database) {
	DOMizeTemplate * DOMizedTemplate = new DOMizeTemplate("../tools_configuration/TOC_Template.html","./"+category+"_toc.html");
	const xml_document& doc = DOMizedTemplate->getDocument();
	this->status = "UNTESTED";

	xpath_node sectionTitle = doc.select_node("//h1[@id='title']");
	if (sectionTitle) {
		this->title = database.getCategoryTitle();
		sectionTitle.node().append_child(node_pcdata).set_value(database.getCategoryTitle().c_str());
	} else {
		std::cout << "section title not found" << std::endl;
	}

	xpath_node sectionDescription = doc.select_node("//p[@id='description']");
	if (sectionDescription) {
		sectionDescription.node().append_child(node_pcdata).set_value(database.getCategoryDescriptionText().c_str());
	} else {
		std::cout << "section description not found" << std::endl;
	}

	xpath_node list = doc.select_node("//ol[@id='list']");
	xml_node firstChildInList = list.node().first_child();
	if (list) {
		for (std::string caseId : database.getIdentificationIds()){
			std::cout << "working on test case: " << caseId << std::endl;
			std::string status = (new TestCase(caseId,category,database))->getStatus();
			xml_node listItem = list.node().append_child("li");
			xml_node anchor = listItem.append_child("a");
			anchor.append_attribute("href") = (category+"/"+caseId+".html").c_str();
			if (status == "PASSED") {
				anchor.append_attribute("style") = "color: 00F000";
			} else if (status == "FAILED") {
				anchor.append_attribute("style") = "color: FF0000";
			}
			anchor.append_child(node_pcdata).set_value((caseId + " - " + database.getIdentificationText(caseId)).c_str());
			updateChapterStatus(status);
		}
	} else {
		std::cout << "list not found" << std::endl;
	}
	DOMizedTemplate->writeDocument();
	delete DOMizedTemplate;

}

Chapter::~Chapter() {
}

const std::string& Chapter::getTitle() const{
	return title;
}

const std::string& Chapter::getStatus() const{
	return status;
}

void Chapter::updateChapterStatus(std::string status) {
	if (status == "UNTESTED") {
		if (this->status != "UNTESTED") {
			this->status = "NOTDONE";
		}
	} else {
		if (status == "FAILED") {
			this->status = "FAILED";
		} else {
			if (this->status == "UNTESTED") {
				this->status = "PASSED";
			}
		}
	}
}
